crates/sui-types/src/transfer.rs

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use move_binary_format::{
+    binary_views::BinaryIndexedView,
+    file_format::{AbilitySet, SignatureToken},
+};
+use move_bytecode_utils::resolve_struct;
+use move_core_types::{
+    account_address::AccountAddress,
+    ident_str,
+    identifier::IdentStr,
+    language_storage::{StructTag, TypeTag},
+};
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    base_types::{ObjectID, SequenceNumber},
+    id::ID,
+    is_object, SUI_FRAMEWORK_ADDRESS,
+};
+
+const TRANSFER_MODULE_NAME: &IdentStr = ident_str!("transfer");
+const INBOX_REFERENCE_STRUCT_NAME: &IdentStr = ident_str!("Receiving");
+
+pub const RESOLVED_RECEIVING_STRUCT: (&AccountAddress, &IdentStr, &IdentStr) = (
+    &SUI_FRAMEWORK_ADDRESS,
+    TRANSFER_MODULE_NAME,
+    INBOX_REFERENCE_STRUCT_NAME,
+);
+
+/// Rust version of the Move sui::transfer::Receiving type
+#[derive(Clone, Serialize, Deserialize, Debug)]
+pub struct Receiving {
+    pub id: ID,
+    pub version: SequenceNumber,
+}
+
+impl Receiving {
+    pub fn new(id: ObjectID, version: SequenceNumber) -> Self {
+        Self {
+            id: ID::new(id),
+            version,
+        }
+    }
+
+    pub fn to_bcs_bytes(&self) -> Vec<u8> {
+        bcs::to_bytes(self).expect("Value representation is owned and should always serialize")
+    }
+
+    pub fn struct_tag() -> StructTag {
+        StructTag {
+            address: SUI_FRAMEWORK_ADDRESS,
+            module: TRANSFER_MODULE_NAME.to_owned(),
+            name: INBOX_REFERENCE_STRUCT_NAME.to_owned(),
+            // TODO: this should really include the type parameters eventually when we add type
+            // parameters to the other polymorphic types like this.
+            type_params: vec![],
+        }
+    }
+
+    pub fn type_tag() -> TypeTag {
+        TypeTag::Struct(Box::new(Self::struct_tag()))
+    }
+
+    pub fn is_valid(
+        view: &BinaryIndexedView<'_>,
+        function_type_args: &[AbilitySet],
+        s: &SignatureToken,
+    ) -> Result<bool, String> {
+        use SignatureToken as S;
+        match s {
+            S::MutableReference(inner) | S::Reference(inner) => {
+                Self::is_valid(view, function_type_args, inner)
+            }
+            S::StructInstantiation(idx, type_args) => {
+                let struct_tag = resolve_struct(view, *idx);
+                Ok(struct_tag == RESOLVED_RECEIVING_STRUCT
+                    && type_args.len() == 1
+                    && is_object(view, function_type_args, &type_args[0])?)
```

Comment:
Do we need this check?

Reply:
We don't since we have the `T: key` constraint. I'll remove